radix-engine/src/blueprints/epoch_manager/epoch_manager.rs

```diff
@@ -383,58 +399,178 @@ impl EpochManagerBlueprint {
             LockFlags::MUTABLE,
         )?;
-        let mut statistic: CurrentProposalStatisticSubstate =
+        let mut statistic_substate: CurrentProposalStatisticSubstate =
             api.field_lock_read_typed(statistic_handle)?;
-        // TODO(emissions): In some next "emissions" PR, capture the concluded epoch's validator
-        // statistics (to be used for unreliability penalty calculation); at the moment we only
-        // reset it.
-        statistic.validator_statistics = (0..next_validator_set.len())
-            .map(|_index| ProposalStatistic::default())
+        let previous_statistics = statistic_substate.validator_statistics;
+
+        // apply emissions
+        Self::emit_validator_rewards(previous_validator_set, previous_statistics, config, api)?;
+
+        // select next validator set
+        let registered_validators: Vec<EpochRegisteredValidatorByStakeEntry> = api
+            .actor_sorted_index_scan_typed(
+                OBJECT_HANDLE_SELF,
+                EPOCH_MANAGER_REGISTERED_VALIDATORS_BY_STAKE_INDEX,
+                config.max_validators,
+            )?;
+        let next_validator_set: BTreeMap<ComponentAddress, Validator> = registered_validators
+            .into_iter()
+            .map(|entry| (entry.component_address, entry.validator))
             .collect();
-        api.field_lock_write_typed(statistic_handle, statistic)?;
-        api.field_lock_release(statistic_handle)?;
 
+        // emit epoch change event
         Runtime::emit_event(
             api,
             EpochChangeEvent {
                 epoch,
-                validators: next_validator_set,
+                validators: next_validator_set.clone(),
             },
         )?;
 
+        // write zeroed statistics of next validators
+        statistic_substate.validator_statistics = (0..next_validator_set.len())
+            .map(|_index| ProposalStatistic::default())
+            .collect();
+        api.field_lock_write_typed(statistic_handle, statistic_substate)?;
+        api.field_lock_release(statistic_handle)?;
+
+        // write next validator set
+        validator_set_substate.validator_set = next_validator_set;
+        api.field_lock_write_typed(validator_set_handle, validator_set_substate)?;
+        api.field_lock_release(validator_set_handle)?;
+
         Ok(())
     }
+
+    /// Emits a configured XRD amount ([`EpochManagerConfigSubstate.total_emission_xrd_per_epoch`])
+    /// and distributes it across the given validator set, according to their stake.
+    fn emit_validator_rewards<Y>(
+        validator_set: BTreeMap<ComponentAddress, Validator>,
+        validator_statistics: Vec<ProposalStatistic>,
+        config: &EpochManagerConfigSubstate,
+        api: &mut Y,
+    ) -> Result<(), RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        let validator_rewards = validator_set
+            .into_iter()
+            .zip(validator_statistics)
```

Comment:
During an epoch, we store validator statistics by validator's numeric index (within the sorted validator set). AFAIR the motivation was to minimize the size of the "round change" transaction (since we may get very long chains of "empty" round changes in some problematic consensus conditions, which I don't even fully understand, so I trust David and Łukasz here).

This means that at some point, "zipping" must happen (either implemented via `.zip()` or via `enumerate() + get(index)` or some equivalent - I tried some options and I wasn't able to improve readability or perceived safety).